File: java/leet_code/2370.java
/**
 * This module is part of a leet code solution. It contains a single class, Solution, which
 * provides a method to find the longest ideal string given a character sequence and an integer.
 * The method uses dynamic programming to solve the problem.
 */
class Solution {

    /**
     * This method calculates the longest ideal string from a given character sequence.
     * An ideal string is defined as a string where the ASCII value difference between any two characters is not more than k.
     *
     * @param s The input character sequence. It is expected to be a non-null CharSequence.
     * @param k The maximum allowed ASCII value difference between any two characters in an ideal string. It is expected to be an integer.
     *
     * @return The length of the longest ideal string that can be formed from the input character sequence. The return type is an integer.
     */
    public static int longestIdealString(final CharSequence s, final int k) {

        // dp array to store the longest ideal string ending at each character in the alphabet
        final int[] dp = new int[27];
        final int n = s.length();

        // Iterate over the character sequence in reverse
        for (int i = n - 1; 0 <= i; i--) {
            final char cc = s.charAt(i);
            // Convert the character to its corresponding index in the dp array
            final int idx = (int) cc - 97;
            int max = Integer.MIN_VALUE;

            // Calculate the range of characters that can be included in the ideal string
            final int left = Math.max((idx - k), 0);
            final int right = Math.min((idx + k), 26);

            // Find the maximum length of an ideal string ending at a character within the range
            for (int j = left; j <= right; j++) {
                max = Math.max(max, dp[j]);
            }

            // Update the dp array with the maximum length of an ideal string ending at the current character
            dp[idx] = max + 1;
        }

        int max = Integer.MIN_VALUE;

        // Find the maximum length of an ideal string from the dp array
        for (final int ele : dp) {
            max = Math.max(ele, max);
        }
        return max;
    }
}

File: csharp/leet_code/2370.cs
/// <summary>
/// This module contains the Solution class which provides a method to find the length of the longest ideal string.
/// </summary>
public class Solution {
    /// <summary>
    /// This method calculates the length of the longest ideal string.
    /// An ideal string is defined as a string where the difference between the ASCII values of any two of its characters is not more than 'k'.
    /// </summary>
    /// <param name="s">The input string. It is expected to be a string of lowercase English alphabets.</param>
    /// <param name="k">The maximum allowed difference between the ASCII values of any two characters in an ideal string. It is expected to be an integer.</param>
    /// <returns>The length of the longest ideal string that can be formed from the input string 's'. The return type is an integer.</returns>
    public int LongestIdealString(string s, int k) {
        // 'len' is an array to store the length of the longest ideal string ending with each character.
        var len = new int[26];
        for (var i = 0; i < s.Length; i++) {
            // 'next' is the length of the longest ideal string ending with the current character.
            var next = 0;
            // For each character in the range [s[i] - k, s[i] + k], update 'next' to be the maximum length of the ideal string ending with that character plus one.
            for (var j = Math.Max(0, s[i] - 'a' - k); j <= Math.Min(25, s[i] - 'a' + k); j++) {
                next = Math.Max(next, len[j] + 1);
            }
            // Update the length of the longest ideal string ending with the current character.
            len[s[i] - 'a'] = next;
        }
        // Return the maximum length among all characters.
        return len.Max();
    }
}
